matrix-bot/src/Network/Matrix/Bot/Sync.hs

```diff
@@ -0,0 +1,103 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RankNTypes #-}
+module Network.Matrix.Bot.Sync ( syncLoop
+                               , getInitialSyncToken
+                               ) where
+
+import Control.Monad.IO.Class ( MonadIO
+                              , liftIO
+                              )
+import Control.Monad.IO.Unlift (MonadUnliftIO)
+import Data.Aeson ( FromJSON (parseJSON)
+                  , ToJSON ( toEncoding
+                           , toJSON
+                           )
+                  , genericParseJSON
+                  , genericToEncoding
+                  , genericToJSON
+                  )
+import Data.Functor ((<&>))
+import qualified Data.Text as T
+import GHC.Generics (Generic)
+import Network.Matrix.Client ( ClientSession
+                             , AccountData(accountDataType)
+                             , EventFilter(efNotTypes)
+                             , Filter(filterAccountData)
+                             , MatrixError( MatrixError
+                                          , meErrcode
+                                          )
+                             , MatrixM
+                             , UserID
+                             , createFilter
+                             , defaultEventFilter
+                             , defaultFilter
+                             , getAccountData
+                             , retry
+                             , setAccountData
+                             , srNextBatch
+                             , syncPoll
+                             )
+
+import Network.Matrix.Bot.ErrorHandling
+import Network.Matrix.Bot.JSON
+import Network.Matrix.Bot.Router
+import Network.Matrix.Bot.State
+
+newtype SyncTokenAccountData = SyncTokenAccountData
+  { stadSyncToken :: T.Text
+  }
+  deriving (Show, Eq, Generic)
+
+instance ToJSON SyncTokenAccountData where
+  toJSON = genericToJSON aesonOptions
+  toEncoding = genericToEncoding aesonOptions
+
+instance FromJSON SyncTokenAccountData where
+  parseJSON = genericParseJSON aesonOptions
+
+instance AccountData SyncTokenAccountData where
+  accountDataType _ = syncTokenAccountDataType
+
+syncTokenAccountDataType :: T.Text
+syncTokenAccountDataType = "invalid.example.fixme.use.real.domain.sync_token"
```

Comment:
Yes, the point is to avoid conflicts. I didn't see it mentioned explicitly for account_data, but generally the spec mandates that custom types used in the matrix API use “java-style” reversed domain names for custom names.

hackage.haskell.org seems to be fine for this cases, so I used `org.haskell.hackage.matrix-bot.sync_token`.